helix-view/src/file_watcher.rs

```diff
@@ -0,0 +1,97 @@
+use crossbeam_channel::unbounded;
+use crossbeam_channel::{select, Receiver};
+use notify::{recommended_watcher, RecommendedWatcher, RecursiveMode, Watcher};
+use std::path::PathBuf;
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver};
+
+pub struct NotifyActor {
+    sender: Sender,
+    watcher: (RecommendedWatcher, Receiver<NotifyEvent>),
+}
+
+impl NotifyActor {
+    pub fn spawn() -> NotifyHandle {
+        let (actor_tx, actor_rx) = crossbeam_channel::unbounded();
+        let (tx, rx) = unbounded_channel();
+        let actor = NotifyActor::new(Box::new(move |msg| tx.send(msg).unwrap()));
+        let thread = jod_thread::Builder::new()
+            .name("FileWatcher".to_string())
+            .spawn(move || actor.run(actor_rx))
+            .expect("Failed to spawn file watcher");
+        NotifyHandle {
+            sender: actor_tx,
+            receiver: rx,
+            thread,
+        }
+    }
+
+    fn next_event(
+        &mut self,
+        receiver: &crossbeam_channel::Receiver<ActorMessage>,
+    ) -> Option<ActorEvent> {
+        let watcher_receiver = &self.watcher.1;
+        select! {
+            recv(receiver) -> it => it.ok().map(ActorEvent::Message),
+            recv(watcher_receiver) -> it => Some(ActorEvent::NotifyEvent(it.unwrap())),
+        }
+    }
+
+    fn new(sender: Sender) -> NotifyActor {
+        let (tx, rx) = unbounded();
+        let watcher: RecommendedWatcher = recommended_watcher(move |e| tx.send(e).unwrap()).unwrap();
+        NotifyActor {
+            sender,
+            watcher: (watcher, rx),
+        }
+    }
+
+    fn run(mut self, inbox: crossbeam_channel::Receiver<ActorMessage>) {
```

Comment:
I think you could merge `next_event` and `run`, you construct an enum then immediately deconstruct it.